colcon_override_check/package_selection/override_check.py

```diff
@@ -0,0 +1,171 @@
+        selected_names = {d.descriptor.name for d in decorators if d.selected}
+
+        # Enumerate the underlay packages.
+        identification_extensions = get_package_identification_extensions()
+        underlay_descriptors = discover_packages(
+            args, identification_extensions)
+        augment_packages(underlay_descriptors)
+        if not underlay_descriptors:
+            return
+
+        # Enumerate workspace packages already built (if applicable).
+        building_or_built = set(selected_names)
+        if hasattr(args, 'install_base'):
+            install_base = Path(args.install_base).resolve()
+
+            # Drop any enumerated underlay packages in our install base.
+            # This could happen when a workspace install base is activated and
+            # then built again.
+            rebuilding = [
+                d for d in underlay_descriptors if d.path == install_base]
+            underlay_descriptors.difference_update(rebuilding)
+            if not underlay_descriptors:
+                return
+            building_or_built.update(d.name for d in rebuilding)
+
+            # Explicitly enumerate which workspace packages have already been
+            # built/installed in case the install base hasn't already been
+            # activated.
+            index_path = install_base / get_relative_package_index_path()
+            if index_path.is_dir():
+                building_or_built.update(p.name for p in index_path.iterdir())
+
+        # Enumerate dependencies from underlay packages to selected packages.
+        reverse_dependencies = {}
+        for d in underlay_descriptors:
+            # If an underlay package is being built or has already been built,
+            # its dependencies are irrelevant because it is already, or will
+            # be, overridden itself.
+            if d.name in building_or_built:
+                continue
+
+            # Consider only runtime dependencies from underlay packages to
+            # selected packages.
+            deps = {
+                d.name for d in d.get_dependencies(categories=('run',))
+                if d.name in selected_names}
+
+            for dep in deps:
+                reverse_dependencies.setdefault(dep, set())
+                reverse_dependencies[dep].add(d.name)
+
+        # Compute the list of paths where potentially overridden packages are
+        # installed to.
+        underlay_paths = {}
+        for d in underlay_descriptors:
+            underlay_paths.setdefault(d.name, [])
+            underlay_paths[d.name].append(str(d.path))
+
+        # Finally, determine what packages are overriding underlay packages
+        overriding_names = selected_names.intersection(underlay_paths)
+
+        # Of those, ignore specifically allowed overrides
+        overriding_names.difference_update(args.allow_overriding)
+
+        # Of those, Ignore overrides which are (effectively) leaf packages
```

Comment:
The way leaf packages are calculated looks fine to me. IIUC leaf packages eliminates the concern about ABI/API incompatibilities in the overridden package, but it doesn't eliminate the include directory search order problem. ros2/ros2#1150 is working to mitigate that for a lot of packages, but colcon doesn't have a way to know if a package has mitigated it. Since it can also lead to undefined behavior at runtime, I would recommend keeping the warning even when overridden packages are leaf packages in the underlay.

Reply:
I'm concerned that we'd be introducing a warning for what could be a valid workflow without giving the developer any way to resolve it properly, forcing them to suppress it. So even if they underlying packages have mitigated the problem, they'd still have to suppress the warning. That doesn't seem like a very good user experience, and will probably cause developers to simply ignore or disable these warnings because they "can't be fixed."

I think we need to come up with a plan for how to detect or declare "safe" packages before we do that, but I'm open to other thoughts.

Reply:
That makes sense. Its unfortunate that it means it won't warn in the example that motivated ros2/ros2#1150 https://github.com/jacobperron/overlay_bug_reproduction , but I suppose all the include directory stuff mitigates that.

Reply:
The fact that it's actually bit us in the wild is a strong case for maintaining some kind of warning for leafy sub-trees like that. I'm having trouble weighing the utility of warning in that case against a boy-who-cried-wolf sort of mentality being established regarding this warning.

Maybe we should disable this part and over-warn until we can come up with a way to detect or declare safe packages? Do we actually have any plans to do that?

Reply:
> Maybe we should disable this part and over-warn until we can come up with a way to detect or declare safe packages? Do we actually have any plans to do that?

I don't think we have any plans. Given the volume of feedback against the first implementation of the warning, I think this PR strikes a good balance.